File: 2022/1월/20220123/서윤혁/주차 요금 계산.py
import math

def solution(fees, records):
    answer = []
    time_cal = {}
    mx_time = 23 * 60 + 59
    
    for x in records:
        x = x.split(' ')
        time = x[0].split(':')
        time = int(time[0]) * 60 + int(time[1])
        
        if x[2] == 'IN':
            if x[1] not in time_cal:
                time_cal[x[1]] = -time
            else:
                time_cal[x[1]] -= time
        else:
            if x[1] not in time_cal:
                time_cal[x[1]] = time
            else:
                time_cal[x[1]] += time
        
    tmp = []
    
    for x, y in time_cal.items():
        if y <= 0:
            y += mx_time
        
        if y <= fees[0]:
            tmp.append((fees[1], int(x)))
        else:       
            tmp.append((fees[1] + math.ceil((y - fees[0]) / fees[2]) * fees[3], int(x)))
    
    tmp.sort(key=lambda x : x[1])
    for x, y in tmp:
        answer.append(x)
    return answer

File: 2022/3월/20220327/안현준/주차요금계산.java
import java.util.*;

class Solution {
    public int getTime(String time) {
        String[] s = time.split(":");
        return Integer.parseInt(s[0]) * 60 + Integer.parseInt(s[1]);
    }
    public int[] solution(int[] fees, String[] records) {
        int[] answer = {};
        Map<String, Integer> in = new TreeMap<>();
        Map<String, Integer> totalTime = new TreeMap<>();

        for (String s : records) {
            String[] arr = s.split(" ");
            if (arr[2].equals("IN")) {
                in.put(arr[1], getTime(arr[0]));
            } else {
                int time = getTime(arr[0]) - in.get(arr[1]);
                totalTime.merge(arr[1], time, Integer::sum);
                in.remove(arr[1]);
            }
        }
        for (String key : in.keySet()) {
            int inTime = in.get(key);
            int outTime = getTime("23:59");

            int time = outTime - inTime;
            totalTime.merge(key, time, Integer::sum);
        }

        answer = new int[totalTime.size()];
        int i = 0;
        for (String key : totalTime.keySet()) {
            int time = totalTime.get(key);
            int value = fees[1];
            if (time > fees[0]) {
                value += fees[3] * (int) Math.ceil((time - fees[0]) / (double) fees[2]);
            }
            answer[i] = value;
            i++;
        }
        return answer;
    }
}